<H1>Economic Lot Sizing</H1>

Consider a discrete-time, finite-horizon inventory control problem where you must satisfy a given demand. $d_t$ for each week $t$. This demand can be satisfied by production in any week $t' \le t$. For each week, there is a per unit production cost, a per unit inventory holding cost, and a setup cost. The setup cost is a fixed cost incurred if any production occurs during the week. 

Since the setup cost is paid at most once each week, regardless of whether you produce 1 unit or 1000, you have an incentive to potentially produce enough during a week to cover the demand for multiple weeks. But, producing in excess of the week's demand requires you to hold inventory. This tradeoff between setup and holding costs can be managed through optimization.

<H2>Initial Formulation</H2>
We'll consider two formulations of the problem. The first has fewer decision variables but has a drawback that we'll soon discover.

<H3>Sets and Indices</H3>
* $t \in \{1,\ldots, T\}$: weeks

<H3>Data</H3>
* $c_t$: per unit cost of production in week $t$
* $h_t$: per unit holding cost for inventory remaining in week $t$
* $f_t$: fixed cost of production in week $t$
* $d_t$: demand for week $t$

<H3>Decision Variables</H3>
* $x_t$: binary variable to indicate whether production occurs in week $t$
* $y_t$: production quantity for week $t$
* $s_t$: inventory carried in to period $t$

<H3>Formulation</H3>
\begin{eqnarray}
\min_{x, y, s} && \sum_{t = 1}^T (f_t x_t + c_t y_t + h_t s_t) \nonumber \\
\mbox{s.t.} && y_t \le (\sum_{t' = t}^{T} d_{t'}) x_t, \;\;t = 1,\ldots, T\nonumber \\
&& s_t = s_{t-1} + y_t - d_t,\;\;t = 1,\ldots,T \nonumber \\
&& s_0 = 0 \nonumber \\
&& x_t \in \{0, 1\},\;\;t = 1,\ldots,T \nonumber \\
&& y_t \ge 0,\;\; t = 1,\ldots,T \nonumber \\
&& s_t \ge 0,\;\; t = 1,\ldots,T \nonumber \\
\end{eqnarray}


In [1]:
demands = [20, 50, 10, 50, 50, 10, 20, 40, 20, 30]

In [2]:
production_costs = [10, 10, 10, 10, 10, 10, 10, 10, 10, 10]

In [3]:
fixed_costs = [100, 100, 100, 100, 100, 100, 100, 100, 100, 100]

In [4]:
holding_costs = [1, 1, 1, 1, 1, 1, 1, 1, 1, 1]

In [5]:
import gurobipy as grb
GRB = grb.GRB

def get_lot_sizes(demands, production_costs, fixed_costs, holding_costs):
    m = grb.Model()
    
    weeks = range(len(demands))
    production_quantity = m.addVars(weeks, name='quantity')
    holding_quantity = m.addVars(weeks, name='hold')
    to_produce = m.addVars(weeks, name='to_produce', vtype=GRB.BINARY)
    m.update()
    
    for week in weeks:
        m.addConstr(production_quantity[week] <= sum(demands[week:])*to_produce[week],
                    name='to_produce_def.' + str(week))
        carried_inventory = holding_quantity[week-1] if week > 0 else 0
        m.addConstr(holding_quantity[week] == carried_inventory + production_quantity[week] - demands[week],
                    name='inventory_flow.' + str(week))

    m.setObjective(production_quantity.prod(production_costs) + to_produce.prod(fixed_costs) + 
                   holding_quantity.prod(holding_costs))

    m.optimize()
    
    if m.Status == GRB.OPTIMAL:
        for week, var in production_quantity.items():
            if var.X > 1e-4:
                print("\U0001f3ed Produce", var.X, "units in Week", week)
    

In [6]:
get_lot_sizes(demands, production_costs, fixed_costs, holding_costs)

Using license file /Users/irisrenteria/gurobi.lic
Academic license - for non-commercial use only
Gurobi Optimizer version 9.0.2 build v9.0.2rc0 (mac64)
Optimize a model with 20 rows, 30 columns and 49 nonzeros
Model fingerprint: 0x00e4c5e5
Variable types: 20 continuous, 10 integer (10 binary)
Coefficient statistics:
  Matrix range     [1e+00, 3e+02]
  Objective range  [1e+00, 1e+02]
  Bounds range     [1e+00, 1e+00]
  RHS range        [1e+01, 5e+01]
Presolve removed 3 rows and 4 columns
Presolve time: 0.00s
Presolved: 17 rows, 26 columns, 42 nonzeros
Variable types: 17 continuous, 9 integer (9 binary)

Root relaxation: objective 3.375304e+03, 21 iterations, 0.00 seconds

    Nodes    |    Current Node    |     Objective Bounds      |     Work
 Expl Unexpl |  Obj  Depth IntInf | Incumbent    BestBd   Gap | It/Node Time

     0     0 3375.30360    0    7          - 3375.30360      -     -    0s
H    0     0                    3930.0000000 3375.30360  14.1%     -    0s
     0     0 3520.1

It is therefore optimal to produce 80 units in week 0 to cover demand through week 2, to produce 130 units in week 3 to cover demand for weeks 3 through 6, and to produce 90 units in week 7 to cover demand for weeks 7 through 9.

Note that the LP relaxation produced a lower bound on the cost of 3375.304, but ultimately that lower bound could not be acheived as the true minimum cost was 3580. It is possible to formulate the problem in a way that the LP relaxation value is tight, and we can obtain the optimal integer solution immediately without resorting to branch and bound.

<H2>A Tighter Formulation</H2>

We'll consider a reformulation of the problem which has quadratically many variables, but that produces an LP relaxation that is tight. While the problem size will increase, this model will scale better since it produces an integer feasible solution immediately upon solving the LP relaxation.

The trick here is to note that production in week $t$ may be used to satisfy demand in any week $t' \ge t$. We can define the production variables to include both a production week $t$ and a demand week $t'$. We'll need to adjust the cost of these variables to include both the variable cost of production and the cost to hold inventory from week $t$ into week $t'$. This adjusted cost is simply $c_t + \sum_{i = t}^{t'-1} h_i$.

<H3>Data</H3>
* $\bar{c}_{tt'}$: cost to produce a unit in week $t$ and hold it as inventory until week $t'$
* $f_t$: fixed cost of production in week $t$

<H3>Decision Variables</H3>
* $x_t$: binary variable to indicate whether production occurs in week $t$
* $y_{tt'}$: number of units produced in week $t$ to satisfy demand in week $t' \ge t$

<H3>Formulation</H3>
\begin{eqnarray}
\min_{x, y} && \sum_{t = 1}^T \sum_{t'=t}^T \bar{c}_{tt'} y_{tt'} + \sum_{t = 1}^T f_t x_t \nonumber \\
\mbox{s.t.} && \sum_{t = 1}^{t'} y_{tt'} = d_{t'},\;\;t' = 1,\ldots, T \nonumber \\
&& y_{tt'} \le d_{t'}x_{t},\;\;t' = 1,\ldots, T,\;\;t=1,\ldots,t' \nonumber \\
&& x_t \in \{0, 1\},\;\;t = 1,\ldots, T \nonumber \\
&& y_{tt'} \ge 0,\;\;t = 1,\ldots, T,\;\;t' = t,\ldots,|T|. \nonumber
\end{eqnarray}

In [7]:
def get_lot_sizes(demands, production_costs, fixed_costs, holding_costs):
    def get_cost(production_week, demand_week):
        return production_costs[production_week] + holding_costs[demand_week]*(demand_week-production_week)
    
    m = grb.Model()
    weeks = range(len(demands))
    costs = {(production_week, demand_week): get_cost(production_week, demand_week)  
            for production_week in weeks
            for demand_week in weeks[production_week:]}                                                                                                  
    
    production_quantity = m.addVars(costs, obj=costs, name="total_cost")

    to_produce = m.addVars(weeks, obj=fixed_costs, name='to_produce', vtype=GRB.BINARY)


    m.update()

    for demand_week in weeks:
        m.addConstr(production_quantity.sum("*", demand_week)==demands[demand_week], name = "demand_constr."+str(demand_week))
    
    for (production_week, demand_week), production_var in production_quantity.items():
        m.addConstr(production_var <= to_produce[production_week]*demands[demand_week], name = "production_constr."+str(production_week)+"."+str(demand_week))
    
    m.optimize()
    
    for (production_week, demand_week), var in production_quantity.items():
        if var.X >= 1e-4:
            print("\U0001f3ed In Week", production_week, "produce", var.X, "units for Week", demand_week)
        
    m.write('lot_sizing.lp')
    m.write('lot_sizing.sol')

In [8]:
get_lot_sizes(demands, production_costs, fixed_costs, holding_costs)

Gurobi Optimizer version 9.0.2 build v9.0.2rc0 (mac64)
Optimize a model with 65 rows, 65 columns and 165 nonzeros
Model fingerprint: 0x77e19905
Variable types: 55 continuous, 10 integer (10 binary)
Coefficient statistics:
  Matrix range     [1e+00, 5e+01]
  Objective range  [1e+01, 1e+02]
  Bounds range     [1e+00, 1e+00]
  RHS range        [1e+01, 5e+01]
Presolve removed 13 rows and 12 columns
Presolve time: 0.00s
Presolved: 52 rows, 53 columns, 132 nonzeros
Variable types: 0 continuous, 53 integer (9 binary)
Found heuristic solution: objective 3820.0000000

Root relaxation: objective 3.580000e+03, 19 iterations, 0.00 seconds

    Nodes    |    Current Node    |     Objective Bounds      |     Work
 Expl Unexpl |  Obj  Depth IntInf | Incumbent    BestBd   Gap | It/Node Time

*    0     0               0    3580.0000000 3580.00000  0.00%     -    0s

Explored 0 nodes (19 simplex iterations) in 0.04 seconds
Thread count was 4 (of 4 available processors)

Solution count 2: 3580 3820 

Op